jps-plugin/src/org/intellij/erlang/jps/builder/ErlangDirtyFilesListsBuilder.java

```diff
@@ -0,0 +1,257 @@
+  private static final String NAME = "Prepare dependencies";
+
+  public ErlangDirtyFilesListsBuilder() {
+    super(Collections.singletonList(ErlangPrepareDirtyFilesListsTargetType.INSTANCE));
+  }
+
+  @Override
+  public void build(@NotNull PrepareErlangDirtyFilesListsTarget target,
+                    @NotNull DirtyFilesHolder<ErlangSourceRootDescriptor, PrepareErlangDirtyFilesListsTarget> holder,
+                    @NotNull BuildOutputConsumer outputConsumer,
+                    @NotNull CompileContext context) throws ProjectBuildException, IOException {
+    LOG.info("Starting prepare dependencies");
+    LOG.debug("Load project build order.");
+    ErlangProjectBuildOrder projectBuildOrder = loadProjectBuildOrder(context);
+    if (projectBuildOrder == null) {
+      addPrepareDependenciesFailedMessage(context);
+      return;
+    }
+
+    LOG.debug("Collect dirty files.");
+    List<String> dirtyErlangFilePaths = new DirtyFileProcessor<String, PrepareErlangDirtyFilesListsTarget>() {
+      @Nullable
+      @Override
+      protected String getDirtyElement(@NotNull File file) throws IOException {
+        return isSourceOrHeader(file) ? file.getAbsolutePath() : null;
+      }
+    }.collectDirtyElements(holder);
+
+    JpsProject project = target.getProject();
+    Map<String, ErlangModuleBuildOrder> moduleBuildOrders;
+    if (dirtyErlangFilePaths.isEmpty()) {
+      LOG.debug("There aren't dirty .erl or .hrl files.");
+      moduleBuildOrders = ContainerUtil.newHashMap();
+    }
+    else {
+      LOG.debug("Search dependent dirty files.");
+      moduleBuildOrders = prepareBuildOrderForDirtyFiles(project, projectBuildOrder, dirtyErlangFilePaths);
+    }
+    LOG.debug("Start writing build orders to XML.");
+    writeBuildOrderToXml(project, context, moduleBuildOrders);
+  }
+
+  @NotNull
+  @Override
+  public String getPresentableName() {
+    return NAME;
+  }
+
+  @Nullable
+  private static ErlangProjectBuildOrder loadProjectBuildOrder(@NotNull CompileContext context) {
+    return readFromXML(BUILD_ORDER_FILE_NAME, context, ErlangProjectBuildOrder.class);
+  }
+
+  private static boolean isSourceOrHeader(@NotNull File file) {
+    return file.getName().endsWith(".erl") || file.getName().endsWith(".hrl");
+  }
+
+  private static Map<String, ErlangModuleBuildOrder> prepareBuildOrderForDirtyFiles(@NotNull JpsProject project,
+                                                                                    @NotNull ErlangProjectBuildOrder dependencyTree,
+                                                                                    @NotNull List<String> dirtyErlangFilePaths) {
+    SortedModuleDependencyGraph graph = new SortedModuleDependencyGraph(dependencyTree);
+    Set<String> allDirtyFiles = getAllDirtyFiles(dirtyErlangFilePaths, GraphGenerator.create(graph));
+    return BuildOrdersCreator.createBuildOrders(project, dependencyTree.myNodes, allDirtyFiles);
```

Comment:
All right, we've already had this problem before. In previous version of this pull request you've had your createBuildOrders method outside of BuildOrdersCreator, and you passed a new instance of BuildOrdersCreator to it.

Now it's the same, but we have two auxiliary classes to implement a single method instead of one.